core/domain/user_services.py

```diff
@@ -2287,3 +2291,272 @@ def get_dashboard_stats(user_id):
+    checkpoints_in_older_exploration,
+    most_distant_reached_checkpoint_state_name_in_older_exploration
+):
+    """Returns the most distant reached checkpoint in current exploration after
+    comparing current exploration with older exploration.
+
+    Args:
+        checkpoints_in_current_exploration: list(str). The checkpoints of
+            current exploration in sequential order.
+        checkpoints_in_older_exploration: list(str). The checkpoints
+            of older exploration in sequential order.
+        most_distant_reached_checkpoint_state_name_in_older_exploration: str.
+            The state name of the most distant reached checkpoint in the older
+            exploration.
+
+    Returns:
+        str or None. The most distant checkpoint in current exploration or
+        None if most distant reached checkpoint of older exploration is not
+        present in current exploration.
+    """
+    # Index of the most_distant_reached_checkpoint in the older exploration.
+    mdrc_index = (
+        checkpoints_in_older_exploration.index(
+            most_distant_reached_checkpoint_state_name_in_older_exploration))
+
+    # Loop through checkpoints of furthest_reached_exploration backwards until
+    # a checkpoint is found that exists in current_exploration too.
+    while mdrc_index >= 0:
+        if checkpoints_in_older_exploration[mdrc_index] in checkpoints_in_current_exploration: # pylint: disable=line-too-long
+            return checkpoints_in_older_exploration[mdrc_index]
+        mdrc_index = mdrc_index - 1
+
+    return None
+
+
+def update_learner_checkpoint_progress(
+    user_id, exploration_id, state_name, exp_version
+):
+    """Sets the furthest reached and most recently reached checkpoint in
+    an exploration by the user.
+
+    Args:
+        user_id: str. The Id of the user.
+        exploration_id: str. The Id of the exploration.
+        state_name: str. The state name of the most recently reached checkpoint.
+        exp_version: int. The exploration version of the most recently reached
+            checkpoint.
+    """
+
+    exploration_user_model = user_models.ExplorationUserDataModel.get(
+        user_id, exploration_id)
+    if exploration_user_model is None:
+        exploration_user_model = user_models.ExplorationUserDataModel.create(
+            user_id, exploration_id)
+
+    current_exploration = exp_fetchers.get_exploration_by_id(
+        exploration_id, True, exp_version)
+
+    # If the exploration is being visited the first time.
+    if exploration_user_model.furthest_reached_checkpoint_state_name is None:
+        exploration_user_model.furthest_reached_checkpoint_exp_version = (
+            exp_version)
+        exploration_user_model.furthest_reached_checkpoint_state_name = (
+            state_name)
+    else:
+        furthest_reached_checkpoint_exploration = (
+            exp_fetchers.get_exploration_by_id(
+                exploration_id,
+                True,
```

Comment:
Should write `strict=True, version=...` here (i.e. always write argnames if they have default values specified). Ditto elsewhere.

@Hudda this should be a lint check, right? Is it implemented/tracked already?

Reply:
The method `get_exploration_by_id` fetches the latest exploration if version is not supplied as a parameter. In this case, I needed to fetch the 'furthest_reached_checkpoint_exploration' and therefore I had to supply that particular exploration version.

Making it strict to always supply parameters even when the method has default values specified would sound a little off in some places. For example: Presently to fetch latest exporation we simply need to use `exp_fetchers.get_exploration_by_id(exp_id)`. We do not know the latest exploration version beforehand since version is a property of exploration dict and we need to fetch the latest exploration to know its version. After making it strict to pass parameters, we would have to call `exp_fetchers.get_exploration_by_id(exp_id, True, None)` with exploration_version paramater supplied as None; which sounds a little absurd. Thoughts?

Reply:
No, I just mean writing this as `exp_fetchers.get_exploration_by_id(exploration_id, strict=True, version=...)` instead of what you have. Existing calls that use the default values can still stay as exp_fetchers.get_exploration_by_id(exp_id).

Reply:
Oh got it now. On it!

Reply:
Done!